File: test/testng/TestSearchTree.java
package testng;

import static org.testng.Assert.*;

import java.util.ArrayList;

import org.testng.annotations.Test;
import utils.data.Cost;
import utils.data.SearchTree;
import utils.data.SearchTreeNode;
import utils.math.MathUtil;

public class TestSearchTree {
    @Test
    public void testGetMinNode() {
        SearchTreeNode<Integer> rn1 = new SearchTreeNode<Integer>(5, new CostOfInteger(5)),
            rn2 = new SearchTreeNode<Integer>(6, new CostOfInteger(6)), rn3 = new SearchTreeNode<Integer>(8, new CostOfInteger(8));
        
        SearchTree<Integer> t1 = new SearchTree<Integer>(rn1),
            t2 = new SearchTree<Integer>(rn2), t3 = new SearchTree<Integer>(rn3);
        
        addNodes(4, 18, rn1);
        addNodes(4, 18, rn2);
        addNodes(4, 18, rn3);

        SearchTreeNode<Integer> oi1 = getMinCostNode(t1),
            oi2 = getMinCostNode(t2), oi3 = getMinCostNode(t3),
            moi1 = getMaxCostNode(t1), moi2 = getMaxCostNode(t2),
            moi3 = getMaxCostNode(t3);

        SearchTreeNode<Integer> ni1 = t1.getMinCostLeaf(),
            ni2 = t2.getMinCostLeaf(), ni3 = t3.getMinCostLeaf(),
            mni1 = t1.getMaxCostLeaf(), mni2 = t2.getMaxCostLeaf(),
            mni3 = t3.getMaxCostLeaf();

        assertEquals(getTotalCost(oi1), getTotalCost(ni1));
        assertEquals(getTotalCost(oi2), getTotalCost(ni2));
        assertEquals(getTotalCost(oi3), getTotalCost(ni3));
        assertEquals(getTotalCost(moi1), getTotalCost(mni1));
        assertEquals(getTotalCost(moi2), getTotalCost(mni2));
        assertEquals(getTotalCost(moi3), getTotalCost(mni3));
        
        assertEquals(oi1.getData(), ni1.getData());
        assertEquals(oi2.getData(), ni2.getData());
        assertEquals(oi3.getData(), ni3.getData());
        assertEquals(moi1.getData(), mni1.getData());
        assertEquals(moi2.getData(), mni2.getData());
        assertEquals(moi3.getData(), mni3.getData());
    }
    
    public void addNodes(int levelToGo, int childNodeCount, SearchTreeNode<Integer> node) {
        if (levelToGo <= 0) {
            return;
        }
        for (int i = 0; i < childNodeCount; i++) {
            int v = MathUtil.randomInteger(0, 10);
            addNodes(levelToGo-1, childNodeCount, node.addChildData(v, new CostOfInteger(v)));
        }
    }
    
    public int getTotalCost(SearchTreeNode<Integer> node) {
        int cost = 0;
        while (node != null) {
            cost += node.getData();
            node = node.getParentNode();
        }
        return cost;
    }
    
    public SearchTreeNode<Integer> getMinCostNode(SearchTree<Integer> tree) {
        int minCost = Integer.MAX_VALUE, minIndex = -1;
        
        ArrayList<SearchTreeNode<Integer>> nodes = tree.getAllLeafNode();
        
        for (int i = 0; i < nodes.size(); i++) {
            int cost = 0;
            SearchTreeNode<Integer> c = nodes.get(i);
            while (c != null) {
                cost += c.getData();
                c = c.getParentNode();
            }
            if (cost < minCost) {
                minCost = cost;
                minIndex = i;
            }
        }
        return nodes.get(minIndex);
    }
    
    public SearchTreeNode<Integer> getMaxCostNode(SearchTree<Integer> tree) {
        int maxCost = Integer.MIN_VALUE, maxIndex = -1;
        
        ArrayList<SearchTreeNode<Integer>> nodes = tree.getAllLeafNode();
        
        for (int i = 0; i < nodes.size(); i++) {
            int cost = 0;
            SearchTreeNode<Integer> c = nodes.get(i);
            while (c != null) {
                cost += c.getData();
                c = c.getParentNode();
            }
            if (cost > maxCost) {
                maxCost = cost;
                maxIndex = i;
            }
        }
        return nodes.get(maxIndex);
    }
    
    private class CostOfInteger implements Cost {
        private int cost;
        
        public CostOfInteger(int c) {
            cost = c;
        }
        
        public int getCost() {
            return cost;
        }

        @Override
        public Cost add(Cost cost) {
            return new CostOfInteger(this.cost + ((CostOfInteger)cost).getCost());
        }

        @Override
        public Cost minus(Cost cost) {
            return new CostOfInteger(this.cost - ((CostOfInteger)cost).getCost());
        }

        @Override
        public int compareTo(Cost o) {
            return cost - ((CostOfInteger)o).getCost();
        }
        
        @Override
        public CostOfInteger clone() {
            return new CostOfInteger(cost);
        }
    }
}

File: test/testng/TestSearchTree.kt
package testng

import org.testng.Assert.*

import java.util.ArrayList

import org.testng.annotations.Test
import utils.data.Cost
import utils.data.SearchTree
import utils.data.SearchTreeNode
import utils.math.MathUtil

class TestSearchTree {
    @Test
    fun testGetMinNode() {
        val rn1 = SearchTreeNode<Int?>(5, CostOfInteger(5))
        val rn2 = SearchTreeNode<Int?>(6, CostOfInteger(6))
        val rn3 = SearchTreeNode<Int?>(8, CostOfInteger(8))

        val t1 = SearchTree<Int?>(rn1)
        val t2 = SearchTree<Int?>(rn2)
        val t3 = SearchTree<Int?>(rn3)

        addNodes(4, 18, rn1)
        addNodes(4, 18, rn2)
        addNodes(4, 18, rn3)

        val oi1 = getMinCostNode(t1)
        val oi2 = getMinCostNode(t2)
        val oi3 = getMinCostNode(t3)
        val moi1 = getMaxCostNode(t1)
        val moi2 = getMaxCostNode(t2)
        val moi3 = getMaxCostNode(t3)

        val ni1 = t1.minCostLeaf
        val ni2 = t2.minCostLeaf
        val ni3 = t3.minCostLeaf
        val mni1 = t1.maxCostLeaf
        val mni2 = t2.maxCostLeaf
        val mni3 = t3.maxCostLeaf

        assertEquals(getTotalCost(oi1), getTotalCost(ni1))
        assertEquals(getTotalCost(oi2), getTotalCost(ni2))
        assertEquals(getTotalCost(oi3), getTotalCost(ni3))
        assertEquals(getTotalCost(moi1), getTotalCost(mni1))
        assertEquals(getTotalCost(moi2), getTotalCost(mni2))
        assertEquals(getTotalCost(moi3), getTotalCost(mni3))

        assertEquals(oi1.getData(), ni1!!.getData())
        assertEquals(oi2.getData(), ni2!!.getData())
        assertEquals(oi3.getData(), ni3!!.getData())
        assertEquals(moi1.getData(), mni1!!.getData())
        assertEquals(moi2.getData(), mni2!!.getData())
        assertEquals(moi3.getData(), mni3!!.getData())
    }

    fun addNodes(levelToGo: Int, childNodeCount: Int, node: SearchTreeNode<Int?>) {
        if (levelToGo <= 0) {
            return
        }
        for (i in 0..childNodeCount - 1) {
            val v = MathUtil.randomInteger(0, 10)
            addNodes(levelToGo - 1, childNodeCount, node.addChildData(v, CostOfInteger(v)))
        }
    }

    fun getTotalCost(node: SearchTreeNode<Int?>?): Int {
        var localNode = node
        var cost = 0
        while (localNode !== null) {
            cost += localNode.getData()!!
            localNode = localNode.getParentNode()
        }
        return cost
    }

    fun getMinCostNode(tree: SearchTree<Int?>): SearchTreeNode<Int?> {
        var minCost = Integer.MAX_VALUE
        var minIndex = -1

        val nodes = tree.getAllLeafNode()!!
        
        for (i in nodes.indices) {
            var cost = 0
            var c: SearchTreeNode<Int?>? = nodes[i]
            while (c !== null) {
                cost += c.getData()!!
                c = c.getParentNode()
            }
            if (cost < minCost) {
                minCost = cost
                minIndex = i
            }
        }
        return nodes[minIndex]
    }

    fun getMaxCostNode(tree: SearchTree<Int?>): SearchTreeNode<Int?> {
        var maxCost = Integer.MIN_VALUE
        var maxIndex = -1

        val nodes = tree.getAllLeafNode()

        for (i in nodes!!.indices) {
            var cost = 0
            var c: SearchTreeNode<Int?>? = nodes[i]
            while (c !== null) {
                cost += c.getData()!!
                c = c.getParentNode()
            }
            if (cost > maxCost) {
                maxCost = cost
                maxIndex = i
            }
        }
        return nodes[maxIndex]
    }

    private inner class CostOfInteger(val cost: Int) : Cost {
        override fun plus(cost: Cost): Cost {
            return CostOfInteger(this.cost + (cost as CostOfInteger).cost)
        }

        override fun minus(cost: Cost): Cost {
            return CostOfInteger(this.cost - (cost as CostOfInteger).cost)
        }

        override fun compareTo(other: Cost): Int {
            return cost - (other as CostOfInteger).cost
        }

        override fun clone(): CostOfInteger {
            return CostOfInteger(cost)
        }
    }
}
